domain/secret/state/state.go

```diff
@@ -1879,7 +1884,150 @@ ON CONFLICT(secret_id) DO UPDATE SET
 		if err := tx.Query(ctx, insertLatestStmt, secret).Run(); err != nil {
 			return errors.Annotatef(err, "updating latest revision %d for cross model secret %q", latestRevision, uri)
 		}
+		if err := st.markObsoleteRevisions(ctx, tx, uri); err != nil {
+			return errors.Annotatef(err, "marking obsolete revisions for secret %q", uri)
+		}
 		return nil
 	})
 	return errors.Trace(domain.CoerceError(err))
 }
+
+// InitialWatchStatementForObsolete returns the initial watch statement and the table name for watching obsolete revisions.
+func (st State) InitialWatchStatementForObsoleteRevision(
+	ctx context.Context, appOwners domainsecret.ApplicationOwners, unitOwners domainsecret.UnitOwners,
+) (string, string) {
+	tableName := "secret_revision"
+
+	// The condition is empty because we want to watch all obsolete revisions.
+	if len(appOwners) == 0 && len(unitOwners) == 0 {
+		return tableName, `
+SELECT uuid
+FROM secret_revision
+WHERE obsolete = true`[1:]
+	}
+
+	if len(unitOwners) > 0 && len(appOwners) > 0 {
+		return tableName, fmt.Sprintf(`
+SELECT sr.uuid
+FROM secret_revision sr
+LEFT JOIN secret_application_owner sao ON sr.secret_id = sao.secret_id
+LEFT JOIN application ON application.uuid = sao.application_uuid
+LEFT JOIN secret_unit_owner suo ON sr.secret_id = suo.secret_id
+LEFT JOIN unit ON unit.uuid = suo.unit_uuid
+WHERE sr.obsolete = true
+AND (
+    sao.application_uuid IS NOT NULL AND application.name IN ('%s')
+    OR suo.unit_uuid IS NOT NULL AND unit.unit_id IN ('%s')
+)`[1:],
+			strings.Join(appOwners, "','"), strings.Join(unitOwners, "','"))
```

Comment:
Agreed, this is because of the current limitation on our event source.
However, I will have a follow-up PR to use the new customizable query feature added last night.